os/kernel/fault_manager/fault_manager.c

```diff
@@ -0,0 +1,285 @@
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
+#include <tinyara/init.h>
+#include <tinyara/board.h>
+#include <debug.h>
+#include <fcntl.h>
+#include "fault_manager.h"
+#ifdef CONFIG_BOARDCTL_RESET
+#include <sys/boardctl.h>
+#include <stdlib.h>
+#endif
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Global Data
+ ****************************************************************************/
+
+/* This structure is used to store the mapping between app id and application path
+ * Its value is filled at a time of registering the application for fault manager
+ */
+struct apptable reg_table[CONFIG_NUM_APPS] = { { -1, " "}, { -1, " "} };
+
+/****************************************************************************
+ * Name: fm_register_app
+ *
+ * Description:
+ *   This function is used to register application for the fault manager.
+ *   This function will map application pid with its application path.
+ *
+ * Input parameters:
+ *   pid    - The pid of the task to be registered
+ *   result - Application path to be registered for the fault manager
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise a -1 (ERROR)  value is returned.
+ *
+ ****************************************************************************/
+int fm_register_app(pid_t pid, const char *path)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_NUM_APPS; i++) {
+		if (!(strcmp(reg_table[i].app_path, path))) {
+			reg_table[i].appid = pid;
+			return OK;
+		}
+	}
+
+	for (i = 0; i < CONFIG_NUM_APPS; i++) {
+		if (reg_table[i].appid == -1) {
+			reg_table[i].appid = pid;
+			strcpy(reg_table[i].app_path, path);
+			return OK;
+			lldbg("ERROR=%d OK =%d\n", ERROR, OK);
+		}
+	}
+
+	svdbg("Total number of apps exceeded\n");
+	return ERROR;
+}
+
+/****************************************************************************
+ * Name: fm_deregister_app
+ *
+ * Description:
+ *   This function is used to deregister application from the fault manager.
+ *   This function will remove mapping of application pid with its application path
+ *   in app table.
+ *   It will be called in case if application is not able to start or it is being
+ *   killed by the fault manager.
+ *
+ * Input parameters:
+ *   pid    - The pid of the task to be deregistered from apptable
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ ****************************************************************************/
+int fm_deregister_app(pid_t pid)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_NUM_APPS; i++) {
+		if (reg_table[i].appid == pid) {
+			reg_table[i].appid = -1;
+			strcpy(reg_table[i].app_path, NULL);
```

Comment:
I think strcpy is not needed, we can just set reg_table[i].app_path to NULL.

Reply:
Modified, please check